pkg/apis/ark/v1/backup_locations.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BackupStorageLocation is a location where Ark stores backup objects.
+type BackupStorageLocation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   BackupStorageLocationSpec   `json:"spec"`
+	Status BackupStorageLocationStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BackupStorageLocationList is a list of BackupStorageLocations.
+type BackupStorageLocationList struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Items             []BackupStorageLocation `json:"items"`
+}
+
+// ObjectStorageLocation specifies the settings necessary to connect to a provider's object storage.
+type ObjectStorageLocation struct {
+	// Provider is the cloud provider where object storage resides.
+	Provider string `json:"provider"`
+
+	// Config is for provider-specific configuration fields.
+	Config map[string]string `json:"config"`
+
+	// Bucket is the bucket to use for object storage.
+	Bucket string `json:"bucket"`
+
+	// Prefix is the path inside a bucket to use for Ark storage. Optional.
+	Prefix string `json:"prefix"`
+}
+
+// BackupStorageLocationSpec defines the specification for an Ark BackupStorageLocation
+type BackupStorageLocationSpec struct {
+	ObjectStorage ObjectStorageLocation `json:"objectStorage"`
```

Comment:
I think this should be a pointer, with an `omitempty` tag - over time we may want to support additional types of backup storage locations, so the idea would be that exactly one would be non-nil.

We might want to do this the same way `PersistentVolumeSource` is implemented - ref. https://github.com/kubernetes/api/blob/master/core/v1/types.go#L298 and https://github.com/kubernetes/api/blob/master/core/v1/types.go#L170

Reply:
Are you thinking of maybe a `StorageType` struct that held pointers to `ObjectStorageLocation` and (for example) `GitStorageLocation`?

Reply:
yeah, exactly

Reply:
And looks like for PVs, `PersistentVolumeSpec` embeds the `PersistentVolumeSource`, so could do the same here